scripts/assign_haplotypes.py

```diff
@@ -0,0 +1,227 @@
+    >>> aa_substitutions_match("", [('HA1', '122D'), ('HA1', '276E')])
+    False
+    >>> aa_substitutions_match("HA1:G78S,HA1:T135A,HA1:R150K,HA1:V223I", [])
+    True
+    >>> aa_substitutions_match("HA1:G78S,HA1:T135A,HA1:R150K,HA1:V223I", [('HA1', '122D'), ('HA1', '276E')])
+    False
+    >>> aa_substitutions_match("HA1:G78S,HA1:T135A,HA1:R150K,HA1:V223I", [('HA1', '135A'), ('HA1', '223I')])
+    True
+
+    """
+    if len(record_substitutions) > 0:
+        record_substitutions = [
+            (sub.split(":")[0], sub.split(":")[1][1:])
+            for sub in record_substitutions.split(",")
+        ]
+    else:
+        record_substitutions = []
+
+    return all(
+        sub in record_substitutions
+        for sub in required_substitutions
+    )
+
+
+def assign_haplotype(record, haplotype_definitions, clade_column, default_haplotype, use_clade_as_default_haplotype=False):
+    """Assign the most precise haplotype to the given record based on the given
+    haplotype definitions.
+
+    >>> haplotype_definitions = {'J.2:135A': {'aa': [('HA1', '135A')], 'clade': 'J.2'}, 'J.3': {'clade': 'J.3'}, 'J.1': {'nuc': ['135C'], 'aa': [('HA1', '123R')]}}
+    >>> assign_haplotype({'subclade': 'J.2', "founderMuts['subclade'].aaSubstitutions": "HA1:A135T"}, haplotype_definitions, "subclade", "unassigned")
+    'unassigned'
+    >>> assign_haplotype({'subclade': 'J.2', "founderMuts['subclade'].aaSubstitutions": "HA1:A135T"}, haplotype_definitions, "subclade", "unassigned", use_clade_as_default_haplotype=True)
+    'J.2'
+    >>> assign_haplotype({'subclade': 'J.2', "founderMuts['subclade'].aaSubstitutions": "HA1:T135A"}, haplotype_definitions, "subclade", "unassigned")
+    'J.2:135A'
+    >>> assign_haplotype({'subclade': 'J', 'aaSubstitutions': 'HA1:K123R,HA1:T135A', 'substitutions': 'A100T,T110C,G135C,T200G'}, haplotype_definitions, "subclade", "unassigned")
+    'J.1'
+    >>> assign_haplotype({'subclade': 'J.3'}, haplotype_definitions, "subclade", "unassigned")
+    'J.3'
+
+    """
+    assigned_name = default_haplotype
+    for name, definition in haplotype_definitions.items():
+        if "clade" in definition:
+            # Try to assign this haplotype based on its clade and clade-specific
+            # substitutions.
+            clade_match = (record[clade_column] == definition["clade"])
+            nucleotide_column = f"founderMuts[\'{clade_column}\'].substitutions"
+            aa_column = f"founderMuts[\'{clade_column}\'].aaSubstitutions"
+        else:
+            # Try to assign this haplotype based on all substitutions.
+            clade_match = True
+            nucleotide_column = "substitutions"
+            aa_column = "aaSubstitutions"
+
+        if (
+            clade_match and
+            nucleotide_substitutions_match(record.get(nucleotide_column, ""), definition.get("nuc", [])) and
+            aa_substitutions_match(record.get(aa_column, ""), definition.get("aa", []))
+        ):
+            assigned_name = name
```

Comment:
This will allow overwriting of the assigned haplotype if multiple matches are in the definitions TSV. This is what we want really - e.g a strain which we want to assign `J.2:8D` will first be assigned `J.2` then overwritten. But it's behaviour [that's tripped us up](https://github.com/nextstrain/augur/issues/1711) in `augur clades`. I don't have a good solution here, other than an understanding that the order of ones definitions TSV is very important.